File: src/main/java/club/sk1er/patcher/tweaker/asm/EnchantmentTransformer.java
package club.sk1er.patcher.tweaker.asm;

import club.sk1er.patcher.tweaker.transform.PatcherTransformer;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.FieldInsnNode;
import org.objectweb.asm.tree.InsnList;
import org.objectweb.asm.tree.InsnNode;
import org.objectweb.asm.tree.JumpInsnNode;
import org.objectweb.asm.tree.LabelNode;
import org.objectweb.asm.tree.MethodInsnNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.VarInsnNode;

public class EnchantmentTransformer implements PatcherTransformer {
  /**
   * The class name that's being transformed
   *
   * @return the class name
   */
  @Override
  public String[] getClassName() {
    return new String[] {"net.minecraft.enchantment.Enchantment"};
  }

  /**
   * Perform any asm in order to transform code
   *
   * @param classNode the transformed class node
   * @param name the transformed class name
   */
  @Override
  public void transform(ClassNode classNode, String name) {
    for (MethodNode methodNode : classNode.methods) {
      String methodName = mapMethodName(classNode, methodNode);

      if (methodName.equals("getTranslatedName")) {
        methodNode.instructions.insertBefore(
            methodNode.instructions.getFirst(), getNumericalName());
      }
    }
  }

  private InsnList getNumericalName() {
    InsnList list = new InsnList();
    list.add(new FieldInsnNode(Opcodes.GETSTATIC, getPatcherConfigClass(), "romanNumerals", "Z"));
    LabelNode labelNode = new LabelNode();
    list.add(new JumpInsnNode(Opcodes.IFEQ, labelNode));
    list.add(new VarInsnNode(Opcodes.ALOAD, 0));
    list.add(new VarInsnNode(Opcodes.ILOAD, 1));
    list.add(
        new MethodInsnNode(
            Opcodes.INVOKESTATIC,
            "club/sk1er/patcher/hooks/EnchantmentHook",
            "getNumericalName",
            "(Lnet/minecraft/enchantment/Enchantment;I)Ljava/lang/String;",
            false));
    list.add(new InsnNode(Opcodes.ARETURN));
    list.add(labelNode);
    return list;
  }
}

File: src/main/kotlin/club/sk1er/patcher/asm/client/EnchantmentTransformer.kt
package club.sk1er.patcher.asm.client

import club.sk1er.patcher.tweaker.transform.PatcherTransformer
import codes.som.anthony.koffee.assembleBlock
import codes.som.anthony.koffee.insns.jvm.*
import org.objectweb.asm.tree.ClassNode

class EnchantmentTransformer : PatcherTransformer {
    override fun getClassName() = arrayOf("net.minecraft.enchantment.Enchantment")

    override fun transform(classNode: ClassNode, name: String) {
        classNode.methods.first {
            val methodName = mapMethodName(classNode, it)
            methodName == "getTranslatedName" || methodName == "func_77316_c"
        }?.apply {
            instructions.insertBefore(instructions.first, `get Numerical Name`())
        }
    }

    // oooooh spooky space in function name ooooh
    private fun `get Numerical Name`() = assembleBlock {
        getstatic("club/sk1er/patcher/config/PatcherConfig", "romanNumerals", boolean)
        ifeq(L["1"])
        new(StringBuilder::class)
        dup
        invokespecial(StringBuilder::class, "<init>", void)
        aload_0
        invokevirtual("net/minecraft/enchantment/Enchantment", "func_77320_a", String::class)
        invokestatic("net/minecraft/util/StatCollector", "func_74838_a", String::class, String::class)
        invokevirtual(StringBuilder::class, "append", StringBuilder::class, String::class)
        ldc(" ")
        invokevirtual(StringBuilder::class, "append", StringBuilder::class, String::class)
        iload_1
        invokevirtual(StringBuilder::class, "append", StringBuilder::class, int)
        invokevirtual(StringBuilder::class, "toString", String::class)
        areturn
        +L["1"]
    }.first
}